algebra/src/fft/domain/mixed_radix_2_domain.rs

```diff
@@ -0,0 +1,409 @@
+            // (like it is in the radix 2 case) so we need to remember which elements we've moved as we go along
+            // and can't use the trick of just swapping when processing the first element of a 2-cycle.
+            //
+            // 2. We need to do q_adicity many merge passes, each of which is a bit more complicated than the
+            // specialized q=2 case.
+
+            // Applying the permutation
+            let mut seen = vec![false; n as usize];
+            for k in 0..n
+                {
+                    let mut i = k as usize;
+                    let mut a_i = a[i];
+                    while !seen[i]
+                        {
+                            let dest = Self::mixed_radix_fft_permute(two_adicity, q_adicity, q as u64, n, i);
+                            let a_dest = a[dest];
+                            a[dest] = a_i;
+
+                            seen[i] = true;
+
+                            a_i = a_dest;
+                            i = dest;
+                        }
+                }
+
+            let omega_q = omega.pow(n_over_q);
+            let mut qth_roots = vec![F::one(); q];
+            for i in 1..q
+                {
+                    qth_roots[i] = qth_roots[i-1] * &omega_q;
+                }
+
+            let mut terms = vec![F::one(); q-1];
+
+            //Doing the q_adicity passes
+            for _ in 0..q_adicity
+                {
+                    let n_over_q_times_m = F::BigInt::from(n/((q*m) as u64));
+                    let w_m = omega.pow(n_over_q_times_m);
+                    let mut k = 0;
+                    while k < (n as usize)
+                        {
+                            let mut w_j = F::one(); // w_j is omega_m ^ j
+                            for j in 0..m
+                                {
+                                    let base_term = a[k+j];
+                                    let mut w_j_i = w_j.clone();
+                                    for i in 1..q
+                                        {
+                                            terms[i - 1] = w_j_i * &a[k+j+(i*m)];
+                                            w_j_i *= &w_j;
+                                        }
+
+                                    for i in 0..q
+                                        {
+                                            a[k+j+(i*m)] = base_term;
+                                            for l in 1..q
+                                                {
+                                                    a[k+j+(i*m)] += &(qth_roots[(i*l)%q] * &terms[l-1]);
+                                                }
+                                        }
+                                    w_j *= &w_m;
+                                }
+                            k += q*m ;
+                        }
+                    m *= q;
+                }
+        }
+        else{
+            #[inline]
```

Comment:
This passage duplicates code. If p_adicity = 0, then one can still apply the permutation as above. Of course, the bit twiddling bitreverse is faster than the algebraic method used in the definition of the permutation. However, I do not see any reason why callling the mixed fft with p_adicity = 0.